[Pages] Functions metrics revision (#7576)
cloudflare/cloudflare-docs: content/pages/platform/functions/metrics.md

weight: 11
---

# Metrics

Functions metrics can help you diagnose issues and understand your workloads by showing performance and usage data for your Functions.

## Functions metrics

Functions metrics aggregate request data for an individual Pages project. To view your Functions metrics:

1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com).
2. In **Account Home**, select **Pages**.
3. In your Pages project, select **Functions Metrics**.

> There are three metrics that can help you understand the health of your Function: 

Reviewer: Personally I would use bullet points here since the order of items doesn't matter.

| Internal error^2 | Workers runtime encountered an error | | internalError | 

1. The Exceeded Resources status may appear when the Worker exceeds a [runtime limit](/workers/platform/limits/#request-limits). The most common cause is excessive CPU time, but is also caused by a script exceeding startup time or free tier limits.

> 2. The Internal Error status may appear when the Workers runtime fails to process a request due to an internal failure in our system. These errors are not caused by any issue with the Function code nor any resource limit. While requests with Internal Error status are rare, some may appear during normal operation. These requests are not counted towards usage for billing purposes. If you notice an elevated rate of requests with Internal Error status, review www.cloudflarestatus.com.

Reviewer: ```suggestion
2. The Internal Error status may appear when the Workers runtime fails to process a request due to an internal failure in our system. These errors are not caused by any issue with the Function code nor any resource limit. While requests with Internal Error status are rare, some may appear during normal operation. These requests are not counted towards usage for billing purposes. If you notice an elevated rate of requests with Internal Error status, check the [Cloudflare Status page](https://www.cloudflarestatus.com/).
```

> To further investigate exceptions, refer to [Debugging and Logging](/pages/platform/functions/debugging-and-logging)

Reviewer: ```suggestion
To further investigate exceptions, refer to [Debugging and Logging](/pages/platform/functions/debugging-and-logging).
```

### CPU time per execution

The CPU Time per execution chart shows historical CPU time data broken down into relevant quantiles using [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling).  Learn more about [interpreting quantiles](https://www.statisticshowto.com/quantile-definition-find-easy-steps/). 

In some cases, higher quantiles may appear to exceed [CPU time limits](/workers/platform/limits/#cpu-runtime) without generating invocation errors because of a mechanism in the Workers runtime that allows rollover CPU time for requests below the CPU limit.

### Duration per execution

The **Duration** chart underneath **Median CPU time** in the **Functions metrics** dashboard shows historical [duration](/workers/platform/limits/#duration) per Function execution. The data is broken down into relevant quantiles, similar to the CPU time chart. 

> Understanding duration on your Function is useful when you are intending to do a significant amount of computation on the Function itself. This is because you may have to use the Unbound usage model which allows up to 30 seconds of CPU time. The Unbound model is charged on duration, refer to [Pricing](/workers/platform/pricing/#workers) for more information. 

Reviewer: ```suggestion
Understanding duration on your Function is useful when you are intending to do a significant amount of computation on the Function itself. This is because you may have to use the Unbound usage model which allows up to 30 seconds of CPU time. The Unbound model is charged on duration. Refer to [Pricing](/workers/platform/pricing/#workers) for more information. 
```
